### ISSUE TITLE
GST / VAT inclusive can't complete sale

### ISSUE THREAD

**Issue by jekkos** [opensourcepos/opensourcepos#1437]
### IMPORTANT IMPORTANT IMPORTANT

1. Chose what you want to report between the sections "New Feature / Enhacement" and "Issue / Question / Bug".
2. Remove just the one of the two sections you don't need 
3. Remove this section as it's just a guideline
4. Fill in all the details as requested by the section you chose

If you remove the complete template when submitting a Bug your issue will be closed as we cannot help without basic information.

### New Feature / Enhacement

*For New Features or Enhacements please clearly write your proposal.
It's important to state whether you expect the community to implement it or you will contribute the work.
Please bear in mind that we will implement new features only on the current code, there is no support for older versions.*

### Issue / Question / Bug

Before submitting an issue please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [x] I'm reporting an issue of an unmodified OSPOS installation
- [x] I checked open and closed issues database and no similar issue was already discussed (please make sure you searched!)
- [x] I read the README, WHATS_NEW and UPGRADE
- [x] I read the FAQ (https://github.com/jekkos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP has php-gd, php-intl, sockets and etc. installed)
- [x] I read the wiki
- [x] I ran any database upgrade scripts (e.g. database/2.4_to_3.0.sql), and migrating function
- [x] I'm aware the latest master could be a development version and therefore not stable
- [x] I know the version of OSPOS and git commit hash (check the footer of your OSPOS), the name and version of OS, Web server, PHP and MySQL and will add them to my issue report

### Installation information

- OSPOS version is: latest master
- PHP version is:   5.6
- MySQL or MariaDB version is:   MySQL 5.5
- OS and version is:   Ubuntu 16.4
- WebServer is:  Apache 2.4

### Expected behaviour
Would be able to complete the sale as there's no amount due (tax included  scenario)

### Actual behaviour
Sale can't be completed for some reason, probably a regression

### Steps to reproduce the issue
![schermafdruk van 2017-08-03 12-06-29](https://user-images.githubusercontent.com/576234/28917025-75eecf8a-7844-11e7-9946-a395c2c70925.png)

**Comment by jekkos** [comment]
Even when I remove the payment manually and increase it above the total the complete button doesn't show

**Comment by daN4cat** [comment]
Is this with the latest master? What's the decimals configuration?

I fixed this issue because Steve removed a logic that I implemented a while back.

**Comment by SteveIreland** [comment]
I have tried to avoided changing this again because of ownership (I was actually "warned off" 😄 ), but my opinion is that whether or not it is "complete" should be based on taxes that are accumulated and eventually written to the `sales_tax` table and which are persisted in currency decimals and therefore should not require any "variance" logic due to rounding.  It was working at one point but subsequent changes apparently broke it and I haven't had a chance to go back in.

However, I would be happy to redo the changes I made if you want.

**Comment by SteveIreland** [comment]
There is also that sub-feature of cash rounding which is tightly coupled to this.

**Comment by jekkos** [comment]
Weird thing is, when I do sale by invoice, I do get the complete button. Why was this mode introduced exactly? If I want to make an invoice, I should now switch to this mode then?

**Comment by jekkos** [comment]
Didn't seem to be a rounding issue as I tried to increase the amount paid but to no avail.. at least I'm able to make an invoice now which is more important ;)

**Comment by jekkos** [comment]
Ok maybe it's not that weird, the complete button on invoice always shows disregarding whether the sale is complete or not. This makes sense I never have my invoices paid before I send them (so having amounts due there), same for quotes I guess.

But what about sales, should we hold on to this logic or not? Should a sale always be paid before it's completed?

**Comment by SteveIreland** [comment]
Maybe just change the "Complete" button to a red background if full payment hasn't been entered?

**Comment by odiea** [comment]
For some reason I do not have that issue with Vat.

**Comment by daN4cat** [comment]
In pure sale case (no invoice) once the payment matches the amount the button changes to complete so people keep pushing the same area. Of course it's completed when the payment transaction is successful, in the case of cash it's just a final confirmation. But I would not change the current behaviour. For invoice case I guess it's different, but then you would need to defer the payment type to a second step. But it's getting complicated in my opinion....

**Comment by daN4cat** [comment]
... or we implement the credit payment logic so an invoice that is not paid at the moment of issue is issue with a credit and later the credit mutates in real payment.
But I guess this needs proper management i.e. a filter in the sales view showing credits and a means to edit and change it for correct payments (although there is already a means to change the payment type when you edit a sale, which I added).

**Comment by SteveIreland** [comment]
I just set up a fresh test installation for both VAT and General Sales Tax and like @odiea I didn't have any issue completing a sale for a VAT sale.

The invoice (at least the variation with my changes) is intended to formalize a sales transaction for which payment has not yet been received and is thus a sale that becomes an accounts receivable transaction (or a debt transaction).  That's why I added the ability to complete a sale with full payment received.

It does what it needs to do so I suppose I don't understand the issue.

If the issue is with your specific test case if you can provide the item price, quantities ordered, any special tax scenario, and any discounts involved then I'd like to test it.  That way we can determine whether or not the "variance" computation is working or if there is something else that is broken.  But so far I can't make it fail.  And I have been very successful today with things failing.

**Comment by daN4cat** [comment]
Shall we close? It has nothing to do with VAT and taxes and neither variance and complete button appearing correctly.
So I understand

**Comment by SteveIreland** [comment]
I haven't yet run through a test on the cash rounding logic,but I can do that today.

**Comment by SteveIreland** [comment]
I was waiting on the Sales controller cleanup before resuming testing.  Will that be merged soon?

**Comment by daN4cat** [comment]
@jekkos shall we close this? Or...

**Comment by jekkos** [comment]
Hmm I'll have another look tomorrow see if I can reproduce..

**Comment by SteveIreland** [comment]
I've tested it a few different ways and cannot recreate it.  I also tested the cash rounding routines and that is now working.

**Comment by esbyrt** [comment]
I just ran into this today. The sale didn't add up correctly and therefore the complete button wasn't showing even though the amounts were correct until I accounted for a few cents missing. Check out the screen shots I've added. The subtotal of $1063.25 minus $106.32 discount should be 956.93 instead of 956.92.  Either way the GST should calculate to $47.85 not $47.88. I apllied an extra payment to get the sale to complete and then when I checked the sale in the daily sales report saw that it's total was actually $1004.81 not the $1004.80 that showed in the Sales Register. Right now I have my tax code set to round up but I am thinking that isn't right. I will have to search your documentation to find out what the rounding methods do.
![sale 1](https://user-images.githubusercontent.com/28318020/33340858-73a16402-d43a-11e7-8471-b900a6ab44f0.JPG)
![sale revised](https://user-images.githubusercontent.com/28318020/33340865-77700e1c-d43a-11e7-8711-6c456f103d16.JPG)
![sale math](https://user-images.githubusercontent.com/28318020/33340870-79eeca16-d43a-11e7-9620-cd54667b1500.JPG)

**Comment by daN4cat** [comment]
Version please?

**Comment by esbyrt** [comment]
Sorry, it's version 3.2.0 - 4f5ad57

**Comment by odiea** [comment]
Have you tried changing either of these settings? I do not know if this applies or not.
$lang["config_cash_decimals_tooltip"] = "If Cash Decimals and Currency Decimals are the same then no cash rounding will take place.

**Comment by odiea** [comment]
Is this a vat issue or  is it a cash rounding issue. I am kinda lost.

**Comment by esbyrt** [comment]
No I haven't tried changing that. I only want 2 decimals in all cases. I'm not sure this applies anyway as the error is there before the payment is applied and the payment method doesn't change anything.

**Comment by daN4cat** [comment]
@SteveIreland you probably need to look into this GST tax calculation issue....

**Comment by SteveIreland** [comment]
I will take a look at it this weekend.  My understanding is that GST tax is a VAT tax.  So the configuration should be "Tax Included" .... right?

**Comment by esbyrt** [comment]
I'm not sure what the definition of VAT tax is but GST is rarely included in the sale amount. If an item is 25.00 an additional 1.25 tax is charged at the checkout for a total of 26.25. Some provinces have provincial tax as well but in Alberta it's just the 5% GST.

**Comment by SteveIreland** [comment]
A pull request has been submitted for this.

**Comment by daN4cat** [comment]
@SteveIreland the issue is not the complete button logic with the variance trick, it's the wrong tax rounding out of a penny with the approximation settings. So I read from @esbyrt report.
You should reproduce the same setup and run the same test to understand what generates that rounding problem.

My complete button variance trick ignores sub set decimal unrounded totals that although the sale register was showing all fine the button was not turning to complete.

**Comment by SteveIreland** [comment]
@daN4cat where do you want to take this from here?  The options appear to be (in my order of preference) 

1. Leave it as is where the variance logic only kicks in only if there is an issue.
2. Simplify the code by removing the variance logic.
3. Give it back to you to sort out.

Excuse my visit down history lane, but I remember that before I made the original change I was able to recreate the issue that the variance logic was intended to account for.  But the resolution I implemented (long ago) was to change how the tax was captured for the purposes of computing the total sale amount - so the variance logic was no longer needed.

There was a something that was reported that triggered you to restore the variance logic but I wasn't involved so I don't know exactly what the data situation was.  I'll check to see if I can find the related issue.

**Comment by esbyrt** [comment]
Hello. I have done some more playing around with the exact same sale I posted pictures of above with the total of $1004.80 that needed a payment of $1004.81. When I entered the sale without any discounts the total came up higher of course but I was able to complete the sale with the displayed total. When I applied a 10% discount to each item as had been there in the original sale it fails with the same problem - the total is out one penny. When I look at the transaction report the subtotal shows as $956.93 NOT the $956.92 that displays on the screen. Now I am wondering if the rounding isn't adding up correctly when the discount is applied? It seems to be the subtotal not the tax that is the issue here. Yes, the GST adds up 3 cents different but that is probably because it is added by line item instead of by invoice total.  This is an ongoing problem so please don't do #1 and just leave it as is.

**Comment by SteveIreland** [comment]
When I said option 1 I was referring to the changes that were included with the pull request that has been waiting on a merge for a couple of days.

Tomorrow I'll retest the proposed change with the example details you've provided (unless I hear otherwise).  My earlier test was with the screen shot you provided further down the thread.

**Comment by daN4cat** [comment]
I'll look into this in the evening and merge the change and see if that fixes the issue here.

**Comment by SteveIreland** [comment]
Thanks for the heads up on that.

**Comment by objecttothis** [comment]
I wouldn't have thought that my rounding change would affect this code, but I may be wrong. If I recall correctly, my change should have only affected the roundup (>= 0.01 gets rounded to 0.10).  Perhaps if the same code is used to round up half cents then it should be reviewed. I'll be available again January 18th, so I can look at it then, or if it needs to get fixed before then someone can rollback that commit and I'll try to come up with a solution for rounding that doesn't break the tax scenario. Alternately someone could fix the code so both work properly.

**Comment by daN4cat** [comment]
@esbyrt what setting did you use for the rounding?

**Comment by daN4cat** [comment]
Fix is in master, complete installation will be available in bintray in 10 - 15 min probably.
However it's just one file change, so @esbyrt if you are fine to take that and change for a test go for it.

**Comment by esbyrt** [comment]
@daN4cat I am using Half Even for rounding. I don't know if that is correct but I haven't found any explanation for what the various rounding settings do.

**Comment by SteveIreland** [comment]
As @esbyrt mentioned there is also an issue in the way discounts are handled in regard to rounding.  I'd like to reopen this issue since I'm working on the fix for that now.  I expect to have a pull request available for this by tomorrow night.

**Comment by SteveIreland** [comment]
Which is preferred for computing discount amount and amount due ... "half round up" or "half round even".

For amount due I'm used to defaulting to "half round up" since there isn't a lot of reason for giving away a penny, but for discounts it might be preferred to "half round down" unless your moral compass for customer service standards is to give the advantage to your customer.  Of course you could go with "half round even" but I've never used that.  It's a bit to random and probably harder to explain to a customer.

**Comment by daN4cat** [comment]
Sorry I closed by mistake.

**Comment by daN4cat** [comment]
Use always round up.

**Comment by SteveIreland** [comment]
Okay.  I've made sure that we are rounding up and have submitted a pull request for it.  It should behave a bit better, but if it doesn't you know where to find me.

**Comment by esbyrt** [comment]
@daN4cat I tried the Fix complete button not showing when paid in full (#1437) code update last night and got a syntax error. I went over the additions line by line and couldn't spot the problem. I left a comment under the code over there. Can you have a look at it and tell me if the problem is in the code or did I miss something? Thanks!

**Comment by odiea** [comment]
The most current is the fix waiting for the pull request.

**Comment by daN4cat** [comment]
Merged

**Comment by daN4cat** [comment]
Apply only the latest change at the moment not the previous one and let us know if it's fine.

**Comment by daN4cat** [comment]
@esbyrt can you please apply the last fix and let us if the situation improved with rounding issue?

**Comment by esbyrt** [comment]
@daN4cat I can't work on it while the store is open so I will try to get to it either tonight or on Sunday and report back. Thanks!

**Comment by daN4cat** [comment]
Ok, please try tonight so we can do something about it over the weekend.

**Comment by esbyrt** [comment]
Okay I tried the code update here https://github.com/opensourcepos/opensourcepos/pull/1744/commits/25809dd2a837443e8e1dc5d6e7f3c9f560e1c2af for the discount rounding issue and on the one sale that has consistently been a problem this fixed it. The total showed as $1004.80 as before and the payment of 1004.80 applied allowed the complete button to show. Looks good!

**Comment by daN4cat** [comment]
Great, so it was a rounding issue. Well done @SteveIreland to fix this issue. Thanks.

**Comment by SteveIreland** [comment]
You're welcome.
